app/webFrontend/src/app/course/course-detail/course-detail.component.ts

```
@@ -39,8 +39,8 @@ export class CourseDetailComponent implements OnInit {
   }
 
   getCourse(courseId: string) {
-    this.courseService.readSingleItem(courseId).then(
-      (course: any) => {
+    this.courseService.readCourseToView(courseId).then(
```

This should be also async and await

---

Hmm, I'm not sure how to elegantly deal with the error handling here when using async / await.
The original function body has 11 LOC _(Just FYI; I don't intend LOC to be a decisive factor.)_.

Await variant 1 ((probably completely) equivalent to the original code), 10 LOC:

```ts
try {
    this.course = await this.courseService.readCourseToView(courseId);
} catch (errorResponse) { // NOTE https://github.com/Microsoft/TypeScript/issues/20024
    if ((<Response>errorResponse).status === 401) {
        this.snackBar.open('You are not authorized to view this course.', '', {duration: 3000});
    }
    return;
}
LastVisitedCourseContainerUpdater.addCourseToLastVisitedCourses(courseId, this.userService, this.userDataService);
this.titleService.setTitleCut(['Course: ', this.course.name]);
```

Await variant 2 (NOT equivalent to the original), 9 LOC:

```ts
try {
    this.course = await this.courseService.readCourseToView(courseId);
    LastVisitedCourseContainerUpdater.addCourseToLastVisitedCourses(courseId, this.userService, this.userDataService);
    this.titleService.setTitleCut(['Course: ', this.course.name]);
} catch (errorResponse) { // NOTE Here errorResponse could be something thrown after the await line!
    if (errorResponse.status === 401) {
        this.snackBar.open('You are not authorized to view this course.', '', {duration: 3000});
    }
}
```

Await variant 3 (functionally equivalent to the original), 8 LOC (+1 with new import):

```ts
// NOTE Requires https://github.com/scopsy/await-to-js (or we could just copy the tiny code)
let errorResponse: Response;
[errorResponse, this.course] = await to(this.courseService.readCourseToView(courseId));
if (!errorResponse) {
    LastVisitedCourseContainerUpdater.addCourseToLastVisitedCourses(courseId, this.userService, this.userDataService);
    this.titleService.setTitleCut(['Course: ', this.course.name]);
} else if (errorResponse.status === 401) {
    this.snackBar.open('You are not authorized to view this course.', '', {duration: 3000});
}
```

So IMO we should either go for variant 3 and add the tiny dependency - or just keep the original.

---

2 and 3 are good solutions, I would do 2, but it is up to you!

---

I would prefer variant 2 too. I think a visible try/catch-Block is much better to see that error handling is handeled at this place and we don't need another dependency here.

---

But variant 2 is functionally flawed, it will capture all exceptions, even those below the `await` line (i.e. anything that might ever be thrown in `addCourseToLastVisitedCourses` or `setTitleCut`). In fact I actually only put it there to demonstrate the issue that refactoring with `await` implies, when error handling is involved.
_(And another minor inconvenience with try/catch in general is the lack of type support for `errorResponse`, at least in the TypeScript version we use.)_
Are absolutely sure about this?

---

As decided in the meeting today, we go for variant 2 with some slight modifications (i.e. `try` line order change and an additional, general `else` error notification to prevent silent failures).